### PR TITLE
feat/SIP-029

### PR DESCRIPTION
**Pull request by jcnelson**
This PR implements the proposed emissions schedule in SIP-029.  It does so by way of introducing `StacksEpochId::Epoch31`, and all of the necessary changes therein (e.g. block-commit memo bump, peer version bump, etc.).

This PR moves the coinbase calculation logic into `StacksEpochId`, and out of `StacksChainState`.  This is done to make epoch-gating slightly easier.  It also adds a means by which unit and integration tests can override the coinbase schedule, which will allow us to test it on a shorter timescale with confidence.

I'm opening this against `develop` because it's backwards-compatible.  The rules don't go into effect until Bitcoin height 875,000.

I'm leaving this as a draft until I can get an integration test in place which tests the new coinbase emissions logic.

### CODE REVIEW

**Comment by will-corcoran** [comment]
This PR will be noted in the "Reference Implementation" section of SIP-029 ([here](https://github.com/stacksgov/sips/pull/196/files))

**Comment by aldur** [comment]
Once merged this will close #5446

File: stackslib/src/chainstate/nakamoto/mod.rs [hunk 1]
```diff
@@ -130,7 +130,7 @@ pub use self::staging_blocks::{
     NakamotoStagingBlocksConn, NakamotoStagingBlocksConnRef, NakamotoStagingBlocksTx,
 };
 
-pub const NAKAMOTO_BLOCK_VERSION: u8 = 0;
+pub const NAKAMOTO_BLOCK_VERSION: u8 = 1;
```
Comment (hstove) [inline stackslib/src/chainstate/nakamoto/mod.rs:133]: Wait, I don't think we want this change, right? This would fail block replay.

**Review by hstove — COMMENTED** [review]
Overall LGTM, including the math I've checked vs the SIP. I have a question about NAKAMOTO_BLOCK_VERSION, though.

File: stackslib/src/chainstate/nakamoto/mod.rs (hunk 1 above)
Comment (hstove) [reply stackslib/src/chainstate/nakamoto/mod.rs:133]: I'm also not sure why this is useful, other than in the `post_blockproposal` check.

Comment (hstove) [reply stackslib/src/chainstate/nakamoto/mod.rs:133]: Gotcha - so there's no issue of old nodes suddenly seeing blocks with a different version? Looking at the code, it doesn't look like it, but I just wanted to double check.

File: stackslib/src/net/api/setloglevel.rs.bak [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+    /// Make the response
+    fn try_handle_request(
+        &mut self,
+        preamble: HttpRequestPreamble,
+        _contents: HttpRequestContents,
+        node: &mut StacksNodeState,
+    ) -> Result<(HttpResponsePreamble, HttpResponseContents), NetError> {
+
+        // TODO
```
Comment (obycode) [inline stackslib/src/net/api/setloglevel.rs.bak:132]: Is this intended to be included yet?

File: testnet/stacks-node/src/tests/nakamoto_integrations.rs [hunk 3]
```diff
@@ -9595,11 +9825,6 @@ fn clarity_cost_spend_down() {
 
     info!("Bootstrapped to Epoch-3.0 boundary, starting nakamoto miner");
 
-    info!("Nakamoto miner started...");
-    blind_signer(&naka_conf, &signers, proposals_submitted);
-
-    wait_for_first_naka_block_commit(60, &commits_submitted);
-
```
[the remark this answers is not in the document]
Comment (jcnelson) [reply testnet/stacks-node/src/tests/nakamoto_integrations.rs:9602]: That's a merge bug. Will fix

File: stackslib/src/net/api/setloglevel.rs.bak (hunk 2 above)
Comment (jcnelson) [reply stackslib/src/net/api/setloglevel.rs.bak:132]: That whole file shouldn't be there. I'll remove it.

File: stackslib/src/net/api/getdiagnostics.rs [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+pub struct RPCDiagnosticsHandler {}
+impl RPCDiagnosticsHandler {
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+/// Node diagnostics
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct RPCDiagnosticsData {
```
[the remark this answers is not in the document]
Comment (jcnelson) [reply stackslib/src/net/api/getdiagnostics.rs:54]: That whole file shouldn't be there. I'll remove it.

File: clarity/src/vm/analysis/type_checker/v2_1/tests/mod.rs [hunk 5]
```diff
@@ -48,6 +48,7 @@ pub mod contracts;
 
 /// Backwards-compatibility shim for type_checker tests. Runs at latest Clarity version.
 pub fn mem_type_check(exp: &str) -> CheckResult<(Option<TypeSignature>, ContractAnalysis)> {
+    // TODO (question for reviewers): This uses Clarity 3 with Epoch 2.1.  Is this a problem?
```
[the remark this answers is not in the document]
Comment (jcnelson) [reply clarity/src/vm/analysis/type_checker/v2_1/tests/mod.rs:51]: Thanks!

File: stackslib/src/chainstate/nakamoto/mod.rs (hunk 1 above)
Comment (jcnelson) [reply stackslib/src/chainstate/nakamoto/mod.rs:133]: Nope, no issue at all.  Block-processing doesn't consider the block version (except for shadow blocks, but even then it only looks at the high bit).

File: stackslib/src/chainstate/stacks/db/blocks.rs [hunk 6]
```diff
@@ -4132,6 +4104,11 @@ impl StacksChainState {
                         current_epoch = StacksEpochId::Epoch30;
                     }
                     StacksEpochId::Epoch30 => {
+                        // no special initialization is needed, since only the coinbase emission
+                        // schedule is changing.
+                        current_epoch = StacksEpochId::Epoch31;
+                    }
+                    StacksEpochId::Epoch31 => {
                         panic!("No defined transition from Epoch30 forward")
```
[the remark this answers is not in the document]
Comment (jcnelson) [reply stackslib/src/chainstate/stacks/db/blocks.rs]: Fixed

File: stacks-common/src/types/mod.rs [hunk 7]
```diff
@@ -89,9 +111,153 @@ pub enum MempoolCollectionBehavior {
...
+}
+
+impl CoinbaseInterval {
+    /// Look up the value of a coinbase at an effective height.
+    /// Precondition: `intervals` must be sorted in ascending order by `effective_start_height`
+    pub fn get_coinbase_at_effective_height(
+        intervals: &[CoinbaseInterval],
+        effective_height: u64,
+    ) -> u128 {
+        if intervals.len() == 0 {
```
[the remark this answers is not in the document]
Comment (jcnelson) [reply stacks-common/src/types/mod.rs]: Fixed

File: clarity/src/vm/analysis/type_checker/v2_1/tests/mod.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (jcnelson) [reply clarity/src/vm/analysis/type_checker/v2_1/tests/mod.rs]: Comment removed

**Review by hstove — APPROVED** [review]

**Review by obycode — APPROVED** [review]
LGTM!

**Review by hstove — APPROVED** [review]
Approving pending CI 👍

**Review by obycode — APPROVED** [review]

**Comment by wileyj** [comment]
@jcnelson having some issues building this locally using latest commit: c99a210c5e22dbd0062704e8de0ce4bf8661c97e

previously, 0177bbfdcb8fc5f71b2ad284b9fa64a98411aa89  built fine (something that came over in the develop merge maybe?)

```
$ cargo build --release --bin stacks-inspect
...
   Compiling stacks-common v0.0.1 (/home/wileyj/stacks-core2/stacks-common)
error[E0004]: non-exhaustive patterns: `&StacksEpochId::Epoch31` not covered
   --> stacks-common/./src/types/mod.rs:346:15
    |
346 |         match self {
    |               ^^^^ pattern `&StacksEpochId::Epoch31` not covered
    |
note: `StacksEpochId` defined here
   --> stacks-common/./src/types/mod.rs:95:10
    |
95  | pub enum StacksEpochId {
    |          ^^^^^^^^^^^^^
...
105 |     Epoch31 = 0x03001,
    |     ------- not covered
    = note: the matched value is of type `&StacksEpochId`
help: ensure that all possible cases are being handled by adding a match arm with a wildcard pattern or an explicit pattern as shown
    |
355 ~             StacksEpochId::Epoch30 => true,
356 ~             &StacksEpochId::Epoch31 => todo!(),
    |

...
```

**Comment by jcnelson** [comment]
Try now

**Comment by wileyj** [comment]
> Try now

confirmed, thanks!

**Review by hstove — APPROVED** [review]

**Review by hstove — APPROVED** [review]

**Review by obycode — APPROVED** [review]

**Pull request merged**